File: doc/haskell/GeometryHigh/TourRotation.hs
-- In Hungarian: Körüljárás (alakzat körüljárása)

module GeometryHigh.TourRotation where

import GeometryLow.Vector
import Logic.LogicalForms
import Number.Infinitesimal
import GeometryLow.Orientation

samplePlusRotTriangle, sampleMinusRotTriangle :: [Point]
samplePlusRotTriangle  = [(0, 0), (2, 0), (1, 1)]
sampleMinusRotTriangle = [(0, 0), (1, 1), (2, 0)]

rotationalEdgeVectors :: [Point] -> [Vector]
rotationalEdgeVectors = map (uncurry fromTo) . tour

rotationalExternalAnglesFloating :: [Point] -> [(Vector, Vector)]
rotationalExternalAnglesFloating = tour. rotationalEdgeVectors

rotationalExternalAngleMeasures :: [Point] -> [Float]
rotationalExternalAngleMeasures = map (uncurry angle2_mP180) . rotationalExternalAnglesFloating

rotationalSumOfExternalAngles :: [Point] -> Float
rotationalSumOfExternalAngles = sum . rotationalExternalAngleMeasures

tourRotDir :: [Point] -> RotationDirection
tourRotDir vertices = let val = rotationalSumOfExternalAngles vertices
                                in if val =~=  360
                                    then PositiveRotation
                                    else if val =~= (-360)
                                        then NegativeRotation
                                        else error $ "Invalid tour rotation value " ++ show val ++ " for polygon vertices tour " ++ show vertices

tourRotDirIsPositive, tourRotDirIsNegative :: [Point] -> Bool
tourRotDirIsPositive = (== PositiveRotation) . tourRotDir
tourRotDirIsNegative = (== NegativeRotation) . tourRotDir

File: app/assets/math/haskell/app/TourRotation.js
/*
-- In Hungarian: Körüljárás (alakzat körüljárása)

module GeometryHigh.TourRotation where

import GeometryLow.Vector
import Logic.LogicalForms
import Number.Infinitesimal
import GeometryLow.Orientation
*/

//samplePlusRotTriangle, sampleMinusRotTriangle :: [Point]
function samplePlusRotTriangle () {return [(0, 0), (2, 0), (1, 1)];}
function sampleMinusRotTriangle() {return [(0, 0), (1, 1), (2, 0)];}

//rotationalEdgeVectors :: [Point] -> [Vector]
function rotationalEdgeVectors(vertices) {return tour(vertices).map(([beginPoint, endPoint]) => fromTo(beginPoint, endPoint));}

//rotationalExternalAnglesFloating :: [Point] -> [(Vector, Vector)]
function rotationalExternalAnglesFloating(vertices) {return tour(rotationalEdgeVectors(vertices));}

//rotationalExternalAngleMeasures :: [Point] -> [Float]
function rotationalExternalAngleMeasures(vertices) {return rotationalExternalAnglesFloating(vertices).map(([a, b]) => angle2_mP180(a, b));}

//rotationalSumOfExternalAngles :: [Point] -> Float
function rotationalSumOfExternalAngles(vertices) {return rotationalExternalAngleMeasures(vertices).reduce((a,b)=>a+b, 0);}

//tourRotDir :: [Point] -> RotationDirection
function tourRotDir(vertices)
{
	var val = rotationalSumOfExternalAngles(vertices);
	switch (true) {
		case ccEq(val,  360): return 'positive-rotation';
		case ccEq(val, -360): return 'negative-rotation';
		default             : throw "Invalid tour rotation value " + val + " for polygon vertices tour " + JSON.stringify(vertices);
	}
}

//tourRotDirIsPositive, tourRotDirIsNegative :: [Point] -> Bool
function tourRotDirIsPositive(vertices) {return tourRotDir(vertices) == 'positive-rotation';}
function tourRotDirIsNegative(vertices) {return tourRotDir(vertices) == 'negative-rotation';}
